ql/src/java/org/apache/hadoop/hive/ql/exec/tez/DagUtils.java

```diff
@@ -1221,9 +1231,20 @@ private Map<String, LocalResource> addTempResources(Configuration conf, String h
         LOG.info("Skipping vertex resource " + file + " that already exists in the session");
         continue;
       }
-      Path hdfsFilePath = new Path(hdfsDirPathStr, getResourceBaseName(new Path(file)));
-      LocalResource localResource = localizeResource(new Path(file),
-          hdfsFilePath, type, conf);
+      Path filePath = new Path(file);
+      Path hdfsFilePath = new Path(hdfsDirPathStr, getResourceBaseName(filePath));
+      String sha = getSha(filePath, conf);
```

Comment:
that wasn't called here before, so we are. adding extra overhead

Reply:
yeah, right, that's a codepath where the duplicated/symlinked hive-exec jars were coming from, so the tradeoff is: calculate an SHA once (in HS2 lifetime) and localize if it's not localized yet vs. localize every jars every time, so it's like:

BEFORE patch
```
SESSION #1
local hive-exec jar: no SHA calc, localize
local hive-exec jar again: no SHA calc, localize
udf1.jar: no SHA calc, localize
udf2.jar: no SHA calc, localize

SESSION #2
local hive-exec jar: no SHA calc, localize
local hive-exec jar again: no SHA calc, localize
udf1.jar: no SHA calc, localize
udf2.jar: no SHA calc, localize
```

AFTER patch
```
local hive-exec jar: SHA calc, localize
local hive-exec jar again: SHA from cache, won't localize again
udf1.jar: SHA calc, localize
udf2.jar: SHA calc, localize

SESSION #2
local hive-exec jar: SHA from cache, localize
local hive-exec jar again: SHA from cache, won't localize again
udf1.jar: SHA calc from cache, localize
udf2.jar: SHA calc from cache, localize
```

so the extra overhead is 1 SHA calculate / each UDF jar / hiveserver2 lifetime (NOT PER SESSION), which is fair enough if you keep using the same UDF jars in subsequent session

Reply:
agree